nutils/types.py

```diff
@@ -1492,6 +1492,56 @@ def del_lru(self):
       lru = self._sorted_keys.pop()
     del self[lru]
 
+def lru_cache(func=None, maxsize=None):
+  '''Buffer-aware LRU cache.
+
+  Returns values from a cache for previously seen arguments. Arguments must be
+  hasheable objects or Numpy arrays, with the caveat that writeable arrays
+  silently disable caching. Arrays are equality-tested based on the underlying
+  buffer, with a callback placed on the memory release that removes the
+  corresponding cache entry.
+  '''
+
+  if func is None:
+    return functools.partial(lru_cache, maxsize=maxsize)
+
+  cache = {} if maxsize is None else lru_dict(maxsize)
+
+  @functools.wraps(func)
+  def wrapped(*args):
+    key = []
+    bases = []
+    for arg in args:
+      if isinstance(arg, numpy.ndarray):
+        obj = arg
+        while True:
+          if obj.flags.writeable:
+            return func(*args)
+          if obj.base is None:
+            break
+          obj = obj.base
+          if isinstance(obj, numpy.lib.stride_tricks.DummyArray):
+            obj = obj.base
+            assert isinstance(obj, numpy.ndarray)
+          elif not isinstance(obj, numpy.ndarray):
+            break
+        bases.append(obj)
+        key.append((Py_buffer(arg).buf, *arg.shape, *arg.strides, arg.dtype.str))
+      else:
+        key.append((type(arg), arg))
+    key = tuple(key)
+    try:
+      v = cache[key]
+    except KeyError:
+      v = cache[key] = func(*args)
+      assert _isimmutable(v)
```

Comment:
What about also add an argument `freeze=None` to `lru_cache` and have 
```python
freeze = lambda x: x if freeze is None else freeze
v = cached[key] = freeze(func(*args))
```
so that the functions in `numeric` can be left untouched. I think it's surprising that they return immutable results even if called with mutable arrays.
It would also be nice to expose an uncached version of `func` with `wrapper.nocache = func`.

Reply:
We actually want to get rid of the polymorphic nature of lru-decorated functions, so in preparation to that it makes more sense to do the work on the function level where it's easier to follow. I added a line in the docstring indicating this future change.